vault/cluster.go

```diff
@@ -342,64 +358,121 @@ func (c *Core) stopClusterListener() {
+			}
 
-	// Prevent data races with the TLS parameters
-	c.clusterParamsLock.Lock()
-	defer c.clusterParamsLock.Unlock()
+			//c.logger.Trace("core: performing cert name lookup", "hello_server_name", clientHello.ServerName, "local_cluster_cert_name", parsedCert.Subject.CommonName)
 
-	forwarding := c.localClusterCert != nil && len(c.localClusterCert) > 0
+			return &tls.Certificate{
+				Certificate: [][]byte{localCert.Bytes()},
+				PrivateKey:  localSigner,
+				Leaf:        parsedCert,
+			}, nil
 
-	var parsedCert *x509.Certificate
-	if forwarding {
-		parsedCert, err = x509.ParseCertificate(c.localClusterCert)
-		if err != nil {
-			return nil, fmt.Errorf("error parsing local cluster certificate: %v", err)
 		}
 
-		// This is idempotent, so be sure it's been added
-		c.clusterCertPool.AddCert(parsedCert)
+		return nil, nil
 	}
 
-	nameLookup := func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	clientLookup := func(requestInfo *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		//c.logger.Trace("core: performing client cert lookup")
+
+		if len(requestInfo.AcceptableCAs) != 1 {
+			return nil, fmt.Errorf("expected only a single acceptable CA")
+		}
+		var localCert bytes.Buffer
+
 		c.clusterParamsLock.RLock()
-		defer c.clusterParamsLock.RUnlock()
+		localCert.Write(c.localClusterCert)
+		localSigner := c.localClusterPrivateKey
+		parsedCert := c.localClusterParsedCert
+		c.clusterParamsLock.RUnlock()
 
-		if forwarding && clientHello.ServerName == parsedCert.Subject.CommonName {
-			return &tls.Certificate{
-				Certificate: [][]byte{c.localClusterCert},
-				PrivateKey:  c.localClusterPrivateKey,
-			}, nil
+		if localCert.Len() == 0 {
+			return nil, fmt.Errorf("forwarding connection client but no local cert")
 		}
 
-		return nil, nil
+		return &tls.Certificate{
+			Certificate: [][]byte{localCert.Bytes()},
+			PrivateKey:  localSigner,
+			Leaf:        parsedCert,
+		}, nil
 	}
 
-	var clientCertificates []tls.Certificate
-	if forwarding {
-		clientCertificates = append(clientCertificates, tls.Certificate{
-			Certificate: [][]byte{c.localClusterCert},
-			PrivateKey:  c.localClusterPrivateKey,
-		})
+	serverConfigLookup := func(clientHello *tls.ClientHelloInfo) (*tls.Config, error) {
+		//c.logger.Trace("core: performing server config lookup")
+		for _, v := range clientHello.SupportedProtos {
+			switch v {
+			case "h2", "req_fw_sb-act_v1":
+			default:
+				return nil, fmt.Errorf("unknown ALPN proto %s", v)
+			}
+		}
+
+		caPool := x509.NewCertPool()
+
+		ret := &tls.Config{
+			ClientAuth:           tls.RequireAndVerifyClientCert,
+			GetCertificate:       serverLookup,
+			GetClientCertificate: clientLookup,
+			MinVersion:           tls.VersionTLS12,
+			RootCAs:              caPool,
+			ClientCAs:            caPool,
+			NextProtos:           clientHello.SupportedProtos,
+		}
+
+		switch {
+		default:
+			c.clusterParamsLock.RLock()
```

Comment:
[Optional]: Can we do this whole block before creating the tlsConfig? Just for readability of code flow.